onnx/defs/data_type_utils.h

```diff
@@ -11,58 +11,109 @@
 #include "onnx/onnx_pb.h"
 
 namespace onnx {
-// String pointer as unique TypeProto identifier.
-typedef const std::string* DataType;
 
-namespace Utils {
+// ONNX domain.
+const char* const ONNX_DOMAIN = "";
+struct DataType;
+
+// DataType pointer as unique TypeProto identifier.
+typedef const DataType* PDataType;
 
+namespace Utils {
 // Data type utility, which maintains a global type string to TypeProto map.
 // DataType (string pointer) is used as unique data type identifier for
 // efficiency.
-//
-// Grammar for data type string:
-// <type> ::= <data_type> |
-//            tensor(<data_type>) |
-//            seq(<type>) |
-//            map(<data_type>, <type>)
-// <data_type> :: = float | int32 | string | bool | uint8
-//                | int8 | uint16 | int16 | int64 | float16 | double
-//
-// NOTE: <type> ::= <data_type> means the data is scalar (zero dimension).
-//
-// Example: float, tensor(float), etc.
-//
 class DataTypeUtils {
  public:
-  static DataType ToType(const std::string& type_str);
+  static void Register(PDataType p_data_type);
+
+  static PDataType ToType(const std::string& type_id);
 
-  static DataType ToType(const TypeProto& type_proto);
+  static PDataType ToType(const TypeProto& type_proto);
 
-  static const TypeProto& ToTypeProto(const DataType& data_type);
+  static std::string GetElementTypeStr(TensorProto_DataType elem_type);
 
  private:
-  static void FromString(const std::string& type_str, TypeProto& type_proto);
+  static std::string GetId(const TypeProto& type_proto);
+
+  static std::unordered_map<std::string, PDataType>& GetTypeIdToDataMap();
+};
+} // namespace Utils
 
-  static void FromDataTypeString(
-      const std::string& type_str,
-      TensorProto::DataType& tensor_data_type);
+struct DataType {
+  virtual ~DataType() {}
+  const std::string& Domain() const;
+  const std::string& Id() const;
+  const std::string& Description() const;
+  const TypeProto& ToProto() const;
 
-  static std::string ToString(
-      const TypeProto& type_proto,
-      const std::string& left = "",
-      const std::string& right = "");
+  static PDataType Tensor_FLOAT;
+  static PDataType Tensor_UINT8;
+  static PDataType Tensor_INT8;
+  static PDataType Tensor_UINT16;
+  static PDataType Tensor_INT16;
+  static PDataType Tensor_INT32;
+  static PDataType Tensor_INT64;
+  static PDataType Tensor_STRING;
+  static PDataType Tensor_BOOL;
+  static PDataType Tensor_FLOAT16;
+  static PDataType Tensor_DOUBLE;
+  static PDataType Tensor_UINT32;
+  static PDataType Tensor_UINT64;
+  static PDataType Tensor_COMPLEX64;
+  static PDataType Tensor_COMPLEX128;
+  static PDataType Map_Int64_String;
```

Comment:
Can we support all kinds of key/value combination?
If user want to use map<string,string>, right now, we don't support. (Some macros may make code clean.)

Reply:
We can, but let's add them as needed.

Reply:
Sounds good.